clang/test/SemaCXX/verbose-trap.cpp

```diff
@@ -0,0 +1,28 @@
+// RUN: %clang_cc1 -std=c++11 -fsyntax-only -fcxx-exceptions -verify %s
+
+#if !__has_builtin(__builtin_verbose_trap)
+#error
+#endif
+
+constexpr char const* constMsg1 = "hello";
+char const* const constMsg2 = "hello";
+char const constMsg3[] = "hello";
+
+template <const char * const str>
+void f(const char * arg) {
+  __builtin_verbose_trap("Argument_must_not_be_null");
```

Comment:
```suggestion
  __builtin_verbose_trap("Arbitrary string literals can be used!");
  __builtin_verbose_trap("Argument_must_not_be_null");
```

I think this would require encoding the message, but it's really important to be able to pass arbitrary string literals here.

Reply:
+1. Can we also check:
1. A non-ASCII Unicode character?
2. A long string (is there any limitation on the string length?)?

Reply:
1. Yes, it can handle something like `__builtin_verbose_trap(u8"¡¢£¤¥¦§¨©ª");`.
2. How long of a string do you need to test? I don't think there's a limitation that's specific to this builtin.

Reply:
Oh, okay, then I don't understand how this works under the hood, but I'm happy :)

Reply:
I noticed that the code doesn't compile in `c++20` because the type of UTF8 string literals is `char8_t[N]` instead of `char[N]`. This is the error message:

`cannot initialize a parameter of type 'const char *' with an lvalue of type 'const char8_t[21]'`

It compiles if I cast it to the correct type (`__builtin_verbose_trap((const char *)u8"¡¢£¤¥¦§¨©ª");`).

How important is it to be able to pass UTF8 string literals to the builtin in c++20 without using any casts?

Reply:
It looks like we'd need another version of the builtin (`__builtin_verbose_trap_8t`?) that accepts `const char8_t *` if we wanted to pass UTF8 string literals in c++20.

Reply:
You can't overload the builtin? IIRC there are others that do this, although I can't name one offhand.

Reply:
FWIW I think not being able to pass actual UTF8 is reasonable. When I said "arbitrary string literal", I meant anything that you can put in a normal `char` string literal, like spaces and special characters.

Reply:
@pogo59,  it looks like it's possible to overload builtins using templates in `clang/Basic/Builtins.td` as you said (for example, see `SyncFetchAndAddN`). However,

1. `char8_t` for builtin types isn't currently supported. We'll need to fix a few places in clang including tablegen to support that type.
2. It seems that the overloaded builtin will need a suffix (e.g., `_u8`) to distinguish it from the original one.

I think we can try to add support for that in a follow up patch unless it's something we need now.

Reply:
You can overload builtins by adding custom type checking to them. The "templates" in `Builtins.td` is more of a way to generate builtins with the same pattern for different types. e.g. there are `__builtin_cos`, `__builtin_cosf` and `__builtin_cosl`, which are essentially identical, except that they take `double`, `float` and `long double` respectively. To avoid typing everything out (and making errors that way), you can make the builtin a template.

Reply:
IMO it's completely reasonable to leave supporting Unicode as a potential follow-up, but I'm happy to know what currently is and isn't supported.

For the string length -- maybe let's try 128 or 256 characters? There isn't any specific size that we need to test -- just want some validation that passing a string longer than a short sentence won't cause any weird effects.

Reply:
@philnik777 you are right. I'll enable unicode in follow-up patches using custom type checking.

@var-const 128 and 256 character strings should work without any problems. I don't think you'll see any problems with long strings unless you hit implementation limits of clang or lldb. I tested a 512 char string literal locally and both clang and lldb work fine.

Reply:
Imposing this length limit seems pretty arbitrary.  What's the actually restriction on the debug info/LLDB side? If there isn't one I don't think we should be adding an artificial limit without a good reason for it.

Reply:
As long as the compiler isn't imposing an artificial limit then that seems fine to me.

Reply:
FWIW, I've seen multiple MB symbol names :) so I probably wouldn't be too worried that any user would write a description that's longer than the symbols that complex templates can create...